Implement G123 gwss method

sanjaynagi opened the pull request:
This PR adds support for the G123 genome-wide selection scan statistic for unphased diplotype data.  I have added a test for each `ag3` and `af1`, and a notebook.

To support G123 calculation, I have made a function `_diplotype_frequencies()` analogous to `_haplotype_frequencies()`. As with the `ihs` functions, I've added support for both a `min` and `max_cohort_size`. 

Main question for @alimanfoo, should I implement a segregating site filter? I think its probably needed, because at the minute we are basically windowing by bases rather than SNPs. 

Resolves #305.

alimanfoo commented:
> Main question for @alimanfoo, should I implement a segregating site filter? I think its probably needed, because at the minute we are basically windowing by bases rather than SNPs

Ideally we'd want some way to calibrate a sensible window size for g123, analogous to h12.

If we window on bases (sites) then the windowing will be unaffected by any genomic variations in diversity within the cohort you're analysing, which is good. But you will get spikes around centromeres because of low diversity generally. Also computation may be slower because you're hashing much more data (lots of nonseg sites).

If we window on segregating sites within the cohort of interest, computation should be more efficient and less prone to peaks in centromeres. But also may reduce sensitivity to detect true selection signals, because low diversity around a signal will cause windows to get stretched. 

Ideally you want to ascertain segregating sites in some much larger cohort first, or in some other cohort where you think it's unlikely they're affected by recent selection. That's effectively what happens with h12, because it's using the haplotypes, and the haplotypes are phased at biallelic sites that we're segregating in Ag3.0.

Hm, tricky...

alimanfoo commented:
One option could be to subset the genotype calls to the same sites that we used for haplotype phasing. This would effectively be using Ag3.0 as a whole to ascertain segregating sites. If we did go down that route, we'd probably want to expose the fact that actually there are three sets of sites to choose from: gamb_colu_arab, gamb_colu and arab.

alimanfoo commented:
Possibly could do something like expose a `sites` parameter to allow different approaches. If value is "all" then use all sites passing site filters, regardless of whether there is any variation. If value is "segregating" then ascertain segregating sites within the cohort that pass site filters. If value is "gamb_colu_arab" or "gamb_colu" or "arab" then use the corresponding sites used for haplotype phasing.

sanjaynagi commented:
I like that idea!

alimanfoo commented:
> I like the idea of using the haplotype sites! and yes will be nice to have the different approaches!

Cool, let's go with that then. Give me a shout if need any help subsetting genotypes to the haplotype positions.

Maybe we need to add some window size calibration functions too, analogous to H12.

sanjaynagi commented:
Ok, Ive now added support for a `sites` parameter which can be either 'all', 'segregating', or one of the phasing analyses. To support the haplotype site retrieval, I've added a method in `anopheles` `_haplotype_sites_for_contig()` which retrieves haplotype sites for a given contig, with a method of the same name in `ag3` which handles joined arms. 

Ive also added a method for G123 calibration which is analogous to H12, and added some tests for that and added it to the notebook. 

A couple of questions:
- There is some code duplication in the gwss and calibration functions, for the code which subsets to either the haplotype positions or finds segregating sites. Is this OK or should I be making a new function which does this?
- The 'all' parameter does not actually do anything (apart from the fact I've added an assert statement into `g123_gwss` to ensure the `sites` parameter is something we expect. Sometimes I run into situations like this, where one of the parameter options is kind of useless, ie, if it wasn't for the assert statement, we could put `sites='gobbledegook'` and it would still run just like I put 'all'. It always feels weird, is there a good approach for handling situations like this? thanks!

or at least, this is what I want to push to the branch, but git is giving me 
`! [remote rejected] g123-gwss-23-03-23 -> g123-gwss-23-03-23 (failure)
error: failed to push some refs to 'git@github.com:sanjaynagi/malariagen-data-python.git'`, possibly due to the github ssh keys changing, trying to figure out.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8337,35 +8351,52 @@ def plot_ihs_gwss(
+        """Compute Garud's G1, G12, G123, G2/G1."""
+        f = _diplotype_frequencies(gt)
+        # convert dict to array of sorted frequencies
+        f = np.sort(np.fromiter(f.values(), dtype=float))[::-1]
```
This seems overly complex...?

alimanfoo replied:
Oh maybe not, forgot the frequencies are coming from a Counter.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8337,35 +8351,52 @@ def plot_ihs_gwss(
+            Contig name (e.g., "2L" or "3RL")
+        window_size : int
+            The size of windows used to calculate g123 over.
+        sites: str, {'all', 'segregating', 'gamb_colu', 'gamb_colu_arab', 'arab'}
+            How to filter sites. 'all' includes all sites that pass
+            site filters, 'segregating' includes only segregating sites for
+            the given cohort, or a phasing analysis identifier can be
+            provided to use sites from the phased haplotype data, which is an
+            approximation to finding segregating sites in the entire ag3/af1 cohort.
```
```suggestion
        sites: str
            How to filter sites. 'all' includes all sites that pass
            site filters, 'segregating' includes only segregating sites for
            the given cohort, or a phasing analysis identifier can be
            provided to use sites from the haplotype data, which is an
            approximation to finding segregating sites in the entire Ag3.0 
            (gambiae complex) or Af1.0 (funestus) cohort.
```

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+        name = self._g123_gwss_cache_name
 
-        debug = self._log.debug
+        assert sites in self.site_mask_ids or sites in [
+            "all",
+            "segregating",
+        ], "sites parameter must be one of 'all', 'segregating', 'gamb_colu', 'gamb_colu_arab', 'arab'"
```
```suggestion
        if sites == DEFAULT:
            sites = self._default_phasing_analysis
        valid_sites = self.phasing_analysis_ids + ("all", "segregating")
        if sites not in valid_sites:
            raise ValueError(f"Invalid value for `sites` parameter, must be one of {valid_sites}.")
```

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
-            xaxis_title="Haplotypes",
-            showlegend=True,
-        )
+        if min_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples < min_cohort_size:
+                raise ValueError(
+                    f"not enough samples ({n_samples}) for minimum cohort size ({min_cohort_size})"
+                )
+        if max_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples > max_cohort_size:
+                rng = np.random.default_rng(seed=random_seed)
+                loc_downsample = rng.choice(
+                    n_samples, size=max_cohort_size, replace=False
+                )
+                loc_downsample.sort()
+                ds_snps = ds_snps.isel(samples=loc_downsample)
```
Suggest to push this logic up into the snp_calls() function (which already has support for the cohort_size parameters).

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            gt = gt.compute()
+        pos = ds_snps["variant_position"].values
+
+        if sites in self.site_mask_ids:
```
```suggestion
        if sites in self.phasing_analysis_ids:
```

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+                inline_array=True,
+                chunks="native",
+            ).compute()
+            hap_site_mask = np.in1d(pos, haplotype_pos)
```
```suggestion
            hap_site_mask = np.in1d(pos, haplotype_pos, assume_unique=True)
```

Apparently can speed up the computation.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            Contig name (e.g., "2L" or "3RL")
+        window_size : int
+            The size of windows used to calculate g123 over.
+        sites: str, {'all', 'segregating', 'gamb_colu', 'gamb_colu_arab', 'arab'}
+            How to filter sites. 'all' includes all sites that pass
+            site filters, 'segregating' includes only segregating sites for
+            the given cohort, or a phasing analysis identifier can be
+            provided to use sites from the phased haplotype data, which is an
+            approximation to finding segregating sites in the entire ag3/af1 cohort.
```
See comment above.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            Contig name (e.g., "2L" or "3RL")
+        window_size : int
+            The size of windows used to calculate g123 over.
+        sites: str, {'all', 'segregating', 'gamb_colu', 'gamb_colu_arab', 'arab'}
+            How to filter sites. 'all' includes all sites that pass
+            site filters, 'segregating' includes only segregating sites for
+            the given cohort, or a phasing analysis identifier can be
+            provided to use sites from the phased haplotype data, which is an
+            approximation to finding segregating sites in the entire ag3/af1 cohort.
```
See comment above.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+    def g123_calibration(
+        self,
+        contig,
+        sites,
```
```suggestion
        sites=DEFAULT,
```

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+        ----------
+        contig: str
+            Contig name (e.g., "2L" or "3RL")
+        sites: str, {'all', 'segregating', 'gamb_colu', 'gamb_colu_arab', 'arab'}
+            How to filter sites. 'all' includes all sites that pass
+            site filters, 'segregating' includes only segregating sites for
+            the given cohort, or a phasing analysis identifier can be
+            provided to use sites from the phased haplotype data, which is an
+            approximation to finding segregating sites in the entire ag3/af1 cohort.
```
See comment above.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            provided to use sites from the phased haplotype data, which is an
+            approximation to finding segregating sites in the entire ag3/af1 cohort.
+        site_mask : str
+            Which site_mask to use. See the `phasing_site_mask_ids`
```
```suggestion
            Which site_mask to use. See the `site_mask_ids`
```

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+
+        # change this name if you ever change the behaviour of this function, to
+        # invalidate any previously cached data
+        name = self._g123_calibration_cache_name
```
Is this defined?

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            random_seed=random_seed,
+        )
+
+        if min_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples < min_cohort_size:
+                raise ValueError(
+                    f"not enough samples ({n_samples}) for minimum cohort size ({min_cohort_size})"
+                )
+        if max_cohort_size is not None:
+            n_samples = ds_snps.dims["samples"]
+            if n_samples > max_cohort_size:
+                rng = np.random.default_rng(seed=random_seed)
+                loc_downsample = rng.choice(
+                    n_samples, size=max_cohort_size, replace=False
+                )
+                loc_downsample.sort()
+                ds_snps = ds_snps.isel(samples=loc_downsample)
```
If push this up into snp_calls() can avoid the duplication here.

alimanfoo commented on malariagen_data/anopheles.py:
```diff
@@ -8374,190 +8405,852 @@ def plot_haplotype_clustering(
+            gt = gt.compute()
+        pos = ds_snps["variant_position"].values
+
+        if sites in self.site_mask_ids:
+            debug("subsetting to haplotype positions")
+            haplotype_pos = self._haplotype_sites_for_contig(
+                contig=contig,
+                analysis=sites,
+                field="POS",
+                inline_array=True,
+                chunks="native",
+            ).compute()
+            hap_site_mask = np.in1d(pos, haplotype_pos)
+            gt = gt.compress(hap_site_mask, axis=0)
+        elif sites == "segregating":
+            debug("subsetting to segregating sites")
+            ac = gt.count_alleles(max_allele=3)
+            seg = ac.is_segregating()
+            gt = gt.compress(seg, axis=0)
```
This is annoying to have to duplicate. Maybe factor out into a common private method, e.g., `_load_gt_for_g123(contig, sites, site_mask)`.

alimanfoo reviewed (commented):
Awesome :cherry_blossom:  

A few small suggestions...

alimanfoo reviewed (approved):
Looks good, thanks @sanjaynagi! :bow:
